2024/incoming/have-an-ai-christmas.pod

```diff
@@ -0,0 +1,282 @@
+Author: lincoln.stein@gmail.com
+Title: Have an AI Christmas!
+Topic: OpenAI::API
+
+=encoding utf8
+
+Ah, for the good old days of yore when every Christmas season our
+families gathered at a coffee table set in front of a roaring hearth,
+our glue, glitter, paintbrushes, pens and card stock spread out in
+front of us, and we lovingly created personal Christmas cards for all
+our friends and relatives.
+
+But who has time for that now? Nowadays, if my family sends out cards
+at all, we buy them in bulk at the local stationery store. Santa would
+not approve!
+
+=head2 AI to the Rescue!
+
+We've already surrendered our coding to Copilot, our research to
+ChatGPT, and some of us are replacing our real-life friends with
+simulated personalities on Character.ai. So why not let AI create our
+Christmas greetings for us?
+
+L<OpenAI::API> to the rescue! This handy-dandy Perl module exposes
+OpenAI's API to Perl, and lets you do chat, text completion, and
+image generation from the command line. What we're going to do is:
+
+=over
+
+=item 1. Get a description of the type of card to generate and (optionally) the name of its recipient.
+
+=item 2. Call OpenAI's chat API to turn the description into an image generation prompt.
+
+=item 3. Call OpenAI's image generation API to generate an image from the prompt and store it online.
+
+=item 4. Use L<Browser::Open> to display the image in your favorite web browser.
+
+=back
+
+When you run the script it does this:
+
+   > perl merry_xmas.pl "A humorous card for my Uncle Jim, who loves cats"
+   The AI elves are hard at work (tippity tappity)...done!
+   
+Your browser will then open up and display something like this:
+
+=for html
+<img src="xmas-card-for-uncle-jim.png" />
+
+=head2 Prequisites
+
+For this script to work, you'll need to install the L<Browser::Open>
+and L<OpenAI::API> modules from CPAN. You'll also need to register an
+OpenAI account, get an API key, and add the key to your environment
+using the name L<OPENAI_API_KEY>. Instructions for this can be found
+at L<https://platform.openai.com/docs/quickstart>.
+
+=head2 The Main Script!
+
+Here's the main script. Save it as merry_xmas.pl:
+
+=begin perl
+
+#!/usr/bin/env perl
+use lib '.','./lib','../lib';
+use strict;
+use Browser::Open 'open_browser';
```

Comment:
I use `Browser::Open` as well. I just found out about https://metacpan.org/pod/Browser::Start in the last month or so. I'm not suggesting you change this, but thought I'd point it out as I thought it was interesting.

Reply:
Thanks for pointing this out. In your opinion, is Browser::Start preferable to Browser::Open? I've frankly only tested Browser::Open on my one linux distribution at home.

Reply:
I haven't tested `Browser::Start` yet, but the author is very responsive and there are no open issues. `Browser::Open` hasn't seen a new release in a long time and there are some open issues: https://rt.cpan.org/Public/Dist/Display.html?Name=Browser-Open I would guess that it may need a new owner at this point.

See also: https://metacpan.org/pod/Browser::Start#SEE-ALSO

It **might** be worth looking at `Browser::Start` for this use case.